speechbrain/utils/metric_stats.py

```diff
@@ -367,6 +370,10 @@ def summarize(self, field=None, threshold=None, beta=1, eps=1e-8):
             a dict with all statistics is returned.
         threshold : float
             If no threshold is provided, equal error rate is used.
+        max_samples: float
```

Comment:
do we really need that to fix the OOM issue in EER computation or is it better to add another argument in EER computation to use a for loop so you trade more memory for less speed ?

Reply:
I am using GTX 2080 with 10G CUDA memory, and limited max samples to 5000. Using the eer/threshold calculated from the 5000 postive + 5000 negative samples. I obtained FAR 0.13936119335841204 and FRR 0.1393614282003676. Not exactly the same, but okay to me. Another way is move postive/negative_socres tensor to cpu and then calculate eer. I agree that there maybe an algorithm trading speed for memory, but I have no clue yet.

Reply:
> do we really need that to fix the OOM issue in EER computation or is it better to add another argument in EER computation to use a for loop so you trade more memory for less speed ?

I am using GTX 2080 with 10G cuda memory, and keep 5000 samples for postive/negative scores. Using the eer/threshold from the eer function, I get FAR 0.13936119335841204 and FRR 0.1393614282003676. Not exactly the same, but the precision is okay to me. Another way is to move postive/negative scores to cpu before calculating eer. I agree there is an algorithm trading speed for memory, but I have no clue how to do it yet.

Reply:
> do we really need that to fix the OOM issue in EER computation or is it better to add another argument in EER computation to use a for loop so you trade more memory for less speed ?

In EER function, I think maybe we can use a for loop to calcuate far for each candidate threshold https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/utils/metric_stats.py#L451.
But it would require some work, and I am not sure how long it would take when postive/negative_scores are a large tensor.

Reply:
What if you just re-use: 

```python
thresholds, _ = torch.sort(torch.cat([positive_scores, negative_scores]))
thresholds = torch.unique(thresholds)

# Adding intermediate thresholds
interm_thresholds = (thresholds[0:-1] + thresholds[1:]) / 2
thresholds, _ = torch.sort(torch.cat([thresholds, interm_thresholds]))
```

does it go OOM ?

Reply:
> https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/utils/metric_stats.py#L451

This thresholds only gives the threshold candidates. We need to calculate all FAR/FRR for all threshold candidates, which is in https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/utils/metric_stats.py#L452 and https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/utils/metric_stats.py#L461. The OOM happens when we try to get a 10000*10000 tensors (for example there are 10000 threshold candidates)

Reply:
> 


This thresholds only gives the threshold candidates. We need to calculate all FAR/FRR for all threshold candidates, which is in https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/utils/metric_stats.py#L452 and https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/utils/metric_stats.py#L461. The OOM happens when we try to get a 10000*10000 tensors (for example there are 10000 threshold candidates)

Reply:
> What if you just re-use:
> 
> ```python
> thresholds, _ = torch.sort(torch.cat([positive_scores, negative_scores]))
> thresholds = torch.unique(thresholds)
> 
> # Adding intermediate thresholds
> interm_thresholds = (thresholds[0:-1] + thresholds[1:]) / 2
> thresholds, _ = torch.sort(torch.cat([thresholds, interm_thresholds]))
> ```
> 
> does it go OOM ?

This thresholds only gives the threshold candidates. We need to calculate all FAR/FRR for all threshold candidates, which is in https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/utils/metric_stats.py#L452 and https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/utils/metric_stats.py#L461. The OOM happens when we try to get a 10000*10000 tensors (for example there are 10000 threshold candidates)

Reply:
my bad you are right. What if you do a for loop over 10k instead of having a matrix of 10k ? 
+ I think that all the metrics should be by default in the cpu wdyt @mravanelli this is a terrible waste of VRAM

Reply:
> my bad you are right. What if you do a for loop over 10k instead of having a matrix of 10k ? Also I think that all the metrics should be by default in the cpu wdyt @mravanelli this is a terrible waste of VRAM. Your solution is acceptable BTW also. We can merge and I can try to use a for loop

Yes use a for loop can significantly reduce the memory usage, and this step is better calculated on cpu. And the tradeoff between memory and time should be carefully studied.